Pill typeahead cleanup

m-e-l-u-h-a-n opened the pull request:
Reference: [https://github.com/zulip/zulip/pull/18592#issuecomment-847948095](https://github.com/zulip/zulip/pull/18592#issuecomment-847948095)
* Add node tests, to test complete working of typeahead functions in `pill_typeahead.js` at one place. Earlier it was only dependent on tests of other widgets that used it and those tests are more focused in testing working of those widgets rather than logic of these functions.
* Remove redundant variables in `pill_typeahead.js`, and use common variables like `include_users` if possible.
**Testing plan:** `node tests`


**GIFs or screenshots:** There shouldn't be any visual changes due to this.

timabbott commented:
@YashRE42 @showell is one of you up for doing a review on the node tests?

(The non-test commit LGTM).

ryanreh99 commented:
Hey @m-e-l-u-h-a-n ,

ok, these changes makes sense to me.
Now that we're splitting the pills to be a set containing 3 types of values, we require to test whether the 2^3 subsets call the correct functions or not.
Could you mention this in the commit message, Initially I was confused as to why we require this due to similar blocks present in `stream_edit` and `settings_user group` test files.

Btw since only type can be processed, I think we should also test the priority.
e.g. - `opts = {stream: true, user_group: true}` The `stream` function gets called and not the function for `user_group`.


Also these lines are common, so could be extracted to a function and the different types of `opts` configurations can be iterated over.

```
input_pill_typeahead_called = false;
pill_typeahead.set_up(fake_input, pill_widget, {user_group: true});
pill_typeahead.set_up(fake_input, pill_widget, opts);
assert(input_pill_typeahead_called);
```

To also test the priorities we could have an expected array where for the example above `stream` would be returned.
_________


Also could you split out the `static/js/pill_typeahead.js` change in a separate commit.
Right now we use `users: true` everywhere, but since we have a check in `source` function and this commit adds one in the `highlighter` function, we should add it to the other 3 functions too.

m-e-l-u-h-a-n commented:
@ryanreh99, updated PR with follwing changes.
* Updated highlighter tests to not always use `fake_html` for asserting.
* Updated `test_matcher` to test user and user_group case more clearly.
* Added `include_user` checks to remaining functions in `pill_typeahead.js`.
* Updated commit messages as suggested to reflect why we need these tests(to test different possible sets of configurations.) despite of having similar tests in `stream_edit` and `settings_user_group`.

ryanreh99 commented on static/js/pill_typeahead.js:
```diff
@@ -85,7 +91,14 @@ export function set_up(input, pills, opts) {
             }
 
             if (include_user_groups) {
-                return group_matcher(query, item) || person_matcher(query, item);
+                if (include_users) {
+                    // Item could be a person here if a pill widget
+                    // supports both person and user groups. So
+                    // checking for both ensures that we don't
+                    // miss relevant matches of any types.
+                    return group_matcher(query, item) || person_matcher(query, item);
+                }
+                return group_matcher(query, item);
```
```suggestion
            let matches = false;
            if (include_user_groups) {
                matches = matches || group_matcher(query, item);
            }
            if (include_users) {
                matches = matches || person_matcher(query, item);
            }
            return matches;
```

ryanreh99 commented on static/js/pill_typeahead.js:
```diff
@@ -70,9 +70,15 @@ export function set_up(input, pills, opts) {
 
             if (include_user_groups) {
-                return typeahead_helper.render_person_or_user_group(item);
+                if (include_users) {
+                    return typeahead_helper.render_person_or_user_group(item);
+                }
+                return typeahead_helper.render_user_group(item);
             }
-
+            // After reaching this point, it is sure
+            // that given item is a person. So we do
+            // not need include_users check at this
+            // point.
             return typeahead_helper.render_person(item);
```
```suggestion
            if (include_user_groups && user_groups.is_user_group(item)) {
                return typeahead_helper.render_user_group(item);
            }
            
            // Acts as a `include_users` case as well as the default case.
            return typeahead_helper.render_person(item);
```

ryanreh99 reviewed (approved):
LGTM 👍

Just added 2 suggestions.
And I guess it would be better to have `person_matcher` and `group_matcher`'s return be `false` instead of `undefined` ?

m-e-l-u-h-a-n commented:
I've updated this PR, as suggested.

timabbott commented on static/js/pill_typeahead.js:
```diff
@@ -64,10 +64,17 @@ export function set_up(input, pills, opts) {
+                // if (include_users) {
+                //     return typeahead_helper.render_person_or_user_group(item);
+                // }
+                // return typeahead_helper.render_user_group(item);
```
What's the story with this commented code?

m-e-l-u-h-a-n replied:
Oh, I've removed it now. Usually I do refactoring by commenting older pieces and then writing newer ones to ensure I am not changing the logic unknowingly.

timabbott replied:
Sure, just it's good to do a self-review afterwards to catch these things.

(That said, I prefer to delete the code and see it via `git diff` if need be, since that leaves me with less cleanup to do at the end).

timabbott commented on static/js/pill_typeahead.js:
```diff
@@ -79,19 +86,29 @@ export function set_up(input, pills, opts) {
+                matches = matches || person_matcher(query, item);
+            }
+
+            return matches;
```
I'm not sure this structure makes sense; if we're trying to match a particular item, could we just do:
```
if (include_user_groups && is_user_group(item)) {
    return group_matcher(query, item);
}
```

since we should know what type `item` is?  

(I don't know if we actually have that detail plumbed and available here, but that feels like a more natural way to frame the question)

m-e-l-u-h-a-n replied:
We have details, for writing it as you suggested. But `group_matcher` alraedy does that check, before trying to match the item, and returns `false` if item is not a user group. So, writing it as you suggested will cause us having two checks for same thing one after another.
Current implementation also appears similar to as you suggested but we just check for whether item is a group or not inside `group_matcher` function, this also makes `group_matcher` capable of handling items which are not group .
I haven't changed it for now but if you still feel that it should be, then please mention I would update that.

m-e-l-u-h-a-n commented:
I've rebased this PR to clear conflicts.

timabbott commented:
Merged, thanks @m-e-l-u-h-a-n!
